src/sage/algebras/free_algebra.py

```diff
@@ -467,10 +479,22 @@ def __init__(self, R, n, names):
             raise TypeError("argument R must be a ring")
         self.__ngens = n
         indices = FreeMonoid(n, names=names)
-        cat = AlgebrasWithBasis(R)
+        if degrees is None:
+            cat = AlgebrasWithBasis(R)
+        else:
+            if (len(degrees) == len(names) and
+                all(d in ZZ for d in degrees)):
+                cat = AlgebrasWithBasis(R).Graded()
+            else:
+                raise ValueError("argument degrees must specify an integer for each generator")
+
         CombinatorialFreeModule.__init__(self, R, indices, prefix='F',
                                          category=cat)
         self._assign_names(indices.variable_names())
+        if degrees is None:
+            self._degrees = None
+        else:
+            self._degrees = {g: d for g, d in zip(self.monoid().gens(), degrees)}
```

Comment:
```suggestion
            self._degrees = {g: ZZ(d) for g, d in zip(self.monoid().gens(), degrees)}
```
Let's make sure they really are `ZZ` elements here.